submission/races.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package races
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"reflect"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/ctpolicy"
+)
+
+const (
+	// URLProtocol keeps protocol substring for Log URLs.
+	URLProtocol = "https://"
+
+	// Sleeptim is duration between parallel batch call and subsequent requests to Logs within group.
+	SleepTime = time.Second
+)
+
+type guardedLogs struct {
+	logURLs map[string]bool
+	m       sync.RWMutex
+}
+
+func (gList *guardedLogs) add(logURL string) {
+	gList.m.RLock()
+	gList.logURLs[logURL] = true
+	gList.m.RUnlock()
+}
+
+func (gList *guardedLogs) has(logURL string) bool {
+	gList.m.Lock()
+	exist := gList.logURLs[logURL]
+	gList.m.Unlock()
+	return exist
+}
+
+func (gList *guardedLogs) sumFor(URLs map[string]bool) int {
+	var entries int
+	gList.m.Lock()
+	for url := range URLs {
+		if gList.logURLs[url] {
+			entries++
+		}
+	}
+	gList.m.Unlock()
+	return entries
+}
+
+type result struct {
+	sct    *ct.SignedCertificateTimestamp
+	logURL string
+	err    error
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// sleepTime calculates duration for consequent call. For first paralleStart calls duration is 0, while every next one gets additional dur interval.
+func sleepTime(idx int, parallelStart int, dur time.Duration) time.Duration {
+	return (time.Duration(max(idx+1-parallelStart, 0)) * dur)
```

Comment:
I think this would be easier to follow if the `max` logic were folded in here:
```golang
  if idx < parallelStart {
    return time.Duration(0)
  }
  return time.Duration(idx+1-parallelStart)*dur
```

Reply:
Done.